### PR TITLE
Allow page types to easily restrict what type of requests they respond to

### PR DESCRIPTION
**Pull request by ababic**
The main benefit of these changes is to prevent unwanted cache-bypassing on projects by only permitting GET requests for most pages (POST, PUT and DELETE requests are widely known to bypass edge caching and can put unnecessary strain on servers). From a theoretical perspective, it's also fitting to bring pages a little more in-line with Django views, as both are essentially HTTP response generators, and as such, it makes sense for them to have similar controls.

### CODE REVIEW

File: wagtail/models/__init__.py
Comment (RealOrangeOne) [inline wagtail/models/__init__.py:2098]: Question: Why the property which uppers the methods? This will create new strings for every request, which seems unnecessary.

Django's [`require_http_methods`](https://docs.djangoproject.com/en/5.1/topics/http/decorators/#django.views.decorators.http.require_http_methods) decorator expects methods to be uppercase - I think we can do the same here.

File: wagtail/wagtail_hooks.py
Comment (RealOrangeOne) [inline wagtail/wagtail_hooks.py:558]: Issue: It feels surprising that `OPTIONS` methods aren't handled by the page class at all. Could this be moved into `serve`?

File: wagtail/wagtail_hooks.py
Comment (RealOrangeOne) [inline wagtail/wagtail_hooks.py:563]: Suggestion: Let Django / the app server deal with setting the content length. 
```suggestion
```

File: wagtail/models/__init__.py
Comment (laymonage) [reply wagtail/models/__init__.py:2098]: Probably because Django's [`View.http_method_names`](https://docs.djangoproject.com/en/stable/ref/class-based-views/base/#django.views.generic.base.View.http_method_names) accept lowercase strings...

Not sure if it _only_ accepts lowercase, though.

Comment (ababic) [reply wagtail/models/__init__.py:2098]: @laymonage Your deduction is correct. Would be happy to standardise on upper-case everywhere, so that this isn't needed.

File: wagtail/wagtail_hooks.py
Comment (ababic) [reply wagtail/wagtail_hooks.py:558]: @RealOrange That one gets a little tricky with `RoutablePageMixin`, which overrides `serve()` entirely. Since 'before_serve_page' is called from the view, it would be simpler to do it this way.

Comment (ababic) [reply wagtail/wagtail_hooks.py:558]: @RealOrangeOne Another option here would be to move the logic to separete Page methods, so that `RoutablePageMixin` can easily access / call those from it's implementation of `serve()`

Comment (gasman) [reply wagtail/wagtail_hooks.py:558]: Overriding `serve` is not just a `RoutablePageMixin` thing - it's [a documented recipe](https://docs.wagtail.org/en/stable/reference/pages/model_recipes.html#overriding-the-serve-method), and we routinely tell people that you can put anything in there that you would put in a Django view function. Requiring developers to fulfil certain conditions (such as calling helper methods) in their `serve` methods would be a big step backwards.

Comment (ababic) [reply wagtail/wagtail_hooks.py]: Good point @gasman. So, to best address all this feedback, I've added methods to the `Page` model to allow for overriding, and adjusted the hook to initiate the checking and serving of the OPTIONS response.

**Comment by ababic** [comment]
@RealOrangeOne Any ideas on the Django 4.2 failure? It seems not to like me setting the `ALLOWED` header

File: wagtail/models/__init__.py
Comment (RealOrangeOne) [reply wagtail/models/__init__.py]: Is an enum better? Feels like since Django uses strings everywhere, we should follow suit.

Comment (ababic) [reply wagtail/models/__init__.py]: I feel an enum is better... they are values that already exist in memory, and it's more robust that way... An AttributeError is far more difficult to ignore when you make a typo. It will support uppercase strings too, though

File: wagtail/models/__init__.py
Comment (allcaps) [inline wagtail/models/__init__.py:1408]: Would it be better to only allow GET by default? So that it becomes an opt-in strategy for page types that need to support other methods?

File: wagtail/models/__init__.py
Comment (ababic) [reply wagtail/models/__init__.py]: I would suspect the reason Django and other places use strings is because HTTPMethod wasn't around until Python 10, and the functionality pre-dates that a fair bit, plus Enums weren't quite as easy to work with at that time.

File: wagtail/models/__init__.py
Comment (ababic) [reply wagtail/models/__init__.py:1408]: @allcaps How do you feel about my proposal above? Happy to keep things simple for now?

File: wagtail/models/__init__.py
Comment (ababic) [reply wagtail/models/__init__.py]: @RealOrangeOne How do you feel about the above rationisation? Are you happy to mark this as resolved?

File: wagtail/models/__init__.py
Comment (allcaps) [reply wagtail/models/__init__.py:1408]: Sure, lets keep it simple.

**Review by allcaps — APPROVED** [review]
Thank you @ababic, I think this is a nice addition to Wagtail. So you have my approval.

I won't merge the PR, as I like additional eyes on this.

File: wagtail/models/__init__.py
Comment (lb-) [reply wagtail/models/__init__.py:1408]: I like this idea but agree it's best for a v7 type change maybe. Probably best to raise a new issue after this PR goes in.

Comment (lb-) [reply wagtail/models/__init__.py:1408]: We'll need to consider form pages.

File: docs/reference/models.md
Comment (lb-) [inline docs/reference/models.md:214]: Small docs fix, but I can add this if merging.

```suggestion
        The ``http.HTTPMethod`` enum wasn't added until Python 3.11, so if your project uses an older version of Python, you can use uppercase strings instead. For example:
```

https://docs.python.org/3/library/http.html#http.HTTPMethod

**Comment by lb-** [comment]
This looks great to me and works well, I will rebase, add the docs / release notes, squash some commits and force push.

## When `POST` is not allowed

```py
class FormPage(AbstractEmailForm):
    allowed_http_methods = [HTTPMethod.GET, HTTPMethod.OPTIONS]
```

Screenshot 2024-11-20 at 8 04 12 AM

## When `POST` was added back

```py
class FormPage(AbstractEmailForm):
    allowed_http_methods = [HTTPMethod.GET, HTTPMethod.OPTIONS, HTTPMethod.POST]
```

Screenshot 2024-11-20 at 8 04 35 AM

**Pull request merged**
